hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcServerResponseEncoder.java

```diff
@@ -55,4 +69,33 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
       ctx.write(msg, promise);
     }
   }
+
+  private boolean handleFatalThreshold(ChannelHandlerContext ctx) {
+    int fatalThreshold = rpcServer.getWriteBufferFatalThreshold();
+    if (fatalThreshold <= 0) {
+      return false;
+    }
+
+    Channel channel = ctx.channel();
+    long outboundBytes = NettyUnsafeUtils.getTotalPendingOutboundBytes(channel);
+    if (outboundBytes < fatalThreshold) {
+      return false;
+    }
+
+    NettyServerRpcConnection conn = NettyServerRpcConnection.get(channel);
```

Comment:
Just pass the connection in when creating the encoder? Like what we have done in decoder?

Reply:
The decoder is created after the NettyServerRpcConnection is constructed. The encoder unfortunately is created in NettyRpcServer when the channel is first initialized, before the NettyServerRpcConnection is constructed.  Maybe we can refactor a bit?

Reply:
Let's refactor. Using different pattern for encoder and decoder will make others confusing...

Reply:
Done